16.0 mig commown factor out utm utils

Honeyxilia opened the pull request.

Honeyxilia committed:
Create utm_utils to declare UTM utilitary actions for sources and campaigns

Honeyxilia committed:
Move out code related to UTM from commown

Honeyxilia committed:
Modified action check to verify if a model is abstract

In Odoo 16.0 was added a secondary mixin class in the utm module, utm.source.mixin,
which allows to add only a source to a record without a campaign or medium.

To account for that and potential future mixin classes being added,
we now check if the provided model is abstract to avoid trying to access its attributes.

Honeyxilia committed:
[test] Replaced CRM demo record call by a dummy model record

Beforehand, the module necessitated a dependency to the CRM module, because we used a demo crm.lead in the test,
in order to assert the success of the sources merge.

Since this is the only use, it's better to declare a dummy model which inherits utm.mixin, and to run tests on records from that model.

Honeyxilia committed:
[test] Inserted common operations in setUpClass method

- The same sources are used in both tests, so we define them in the setUpClass method.
- In order for the remove action on source1 to fail in test_action_remove,
  there needs to be a record affiliated with source1, which blocks the deletion.
  So, the dummy record's source is source1 by default.

Honeyxilia committed:
Renamed class and model to more descriptive names

Honeyxilia committed:
[test] Removed Exception try/catch

This try/catch block was only used for debugging purposes in tests.
However, since the raised Exception will stop the execution of the test already, this block is useless.

fcayre commented on utm_utils_merge_and_remove_actions/models/utm.py:
```diff
@@ -11,7 +11,7 @@ def _related_entities_by_field(self):
         for field in fields:
             model = field.model_id.model
-            if model == "utm.mixin":
+            if model in ["utm.mixin", "utm.source.mixin"]:
```
Est-ce qu'on ne peut pas plutôt tester si field.model_id est abstrait ? Ça éviterait ce genre de modification lors de l'ajout d'un nouveau mixin ou quand un mixin change de nom.

Honeyxilia replied:
Bien vu, je vais utiliser l'attribut `_abstract` c:

fcayre commented on utm_utils_merge_and_remove_actions/tests/models.py:
```diff
@@ -0,0 +1,6 @@
+
+class TestUtmDummy(models.Model):
+    _name = "test.utm_dummy"
+    _inherit = "utm.mixin"
```
Est-ce que du coup il faudrait pas le faire hériter du mixin qui se trouve dans odoo ?

fcayre replied:
Message de commit : "dummy module" > "dummy model"

Honeyxilia replied:
Je ne comprends pas la question de l'héritage du mixin, étant donné que `utm.mixin` est celui qui est déclaré dans le code d'Odoo (`odoo/addons/utm/models/utm_mixin.py`)

fcayre commented on utm_utils_merge_and_remove_actions/tests/test_utm_source.py:
```diff
@@ -6,33 +6,34 @@
+        super().setUpClass()
+
+        cls.source1 = cls.env.ref("utm.utm_source_search_engine")
+        cls.source2 = cls.source1.copy({"name": "test"})
```
Please move these lines back after the FakeModelLoader dance to minimise the diff.

Honeyxilia replied:
fait c:

fcayre commented on utm_utils_merge_and_remove_actions/tests/test_utm_source.py:
```diff
@@ -35,5 +35,5 @@ def test_action_remove(self):
             self.source2.action_remove()
-        except Exception as exc:
+        except Exception as exc:  # pragma: nocover
             self.fail("Source removal raised '%s' unexpectedly" % exc)
```
Simply remove the try... except, useless IMHO.

Honeyxilia replied:
fait c:

Honeyxilia committed:
[test] Added checks for source removal

In order to clarify the goal of the actions, we added checks to assert that
the merged/removed source was indeed deleted.

Honeyxilia commented:
/ocabot merge patch
